WalletWasabi/Wallets/FileSystemBlocksRepository.cs

```diff
@@ -0,0 +1,133 @@
+using NBitcoin;
+using NBitcoin.DataEncoders;
+using Nito.AsyncEx;
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets
+{
+	/// <summary>
+	/// FileSystemBlocksRepository is a blocks repository that keep the blocks in the file system.
+	/// </summary>
+	public class FileSystemBlocksRepository : IRepository<uint256, Block>
+	{
+		public FileSystemBlocksRepository(string blocksFolderPath, Network network)
+		{
+			BlocksFolderPath = blocksFolderPath;
+			Network = network;
+			CreateFolders();
+		}
+
+		public string BlocksFolderPath { get; }
+		public Network Network { get; }
+		private AsyncLock BlockFolderLock { get; } = new AsyncLock();
+
+		/// <summary>
+		/// Gets a bitcoin block from the file system.
+		/// </summary>
+		/// <param name="hash">The block's hash that identifies the requested block.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		/// <returns>The requested bitcoin block.</returns>
+		public async Task<Block> GetAsync(uint256 hash, CancellationToken cancellationToken)
+		{
+			// Try get the block
+			Block block = null;
+			using (await BlockFolderLock.LockAsync())
+			{
+				var encoder = new HexEncoder();
+				var filePath = Path.Combine(BlocksFolderPath, hash.ToString());
+				if (File.Exists(filePath))
+				{
+					try
+					{
+						var blockBytes = await File.ReadAllBytesAsync(filePath, cancellationToken);
+						block = Block.Load(blockBytes, Network);
+					}
+					catch
+					{
+						// In case the block file is corrupted and we get an EndOfStreamException exception
+						// Ignore any error and continue to re-downloading the block.
+						Logger.LogDebug($"Block {hash} file corrupted, deleting file and block will be re-downloaded.");
+						File.Delete(filePath);
+					}
+				}
+			}
+
+			return block;
+		}
+
+		/// <summary>
+		/// Saves a bitcoin block in the file system.
+		/// </summary>
+		/// <param name="block">The block to be persisted in the file system.</param>
+		/// <param name="cancellationToken">The cancellation token.</param>
+		public async Task SaveAsync(Block block, CancellationToken cancellationToken)
+		{
+			var path = Path.Combine(BlocksFolderPath, block.GetHash().ToString());
```

Comment:
Just a question, why was this declared outside the `using`, [here](https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/Wallets/Wallet.cs#L683-L688) it was inside?

Reply:
I will revert that.